File: src/main/kotlin/ru/xpendence/slimer/mapper/impl/UserParamsMapper.java
package ru.xpendence.slimer.mapper.impl;

import org.springframework.stereotype.Component;
import ru.xpendence.slimer.dto.UserParamsDto;
import ru.xpendence.slimer.entity.UserParams;
import ru.xpendence.slimer.mapper.AbstractMapper;
import ru.xpendence.slimer.mapper.Mapper;
import ru.xpendence.slimer.repository.UserRepository;

import javax.annotation.PostConstruct;

/**
 * Author: Vyacheslav Chernyshov
 * Date: 2019-06-13
 * Time: 18:47
 * e-mail: 2262288@gmail.com
 */
@Component
@Mapper(entity = UserParams.class, dto = UserParamsDto.class)
public class UserParamsMapper extends AbstractMapper<UserParams, UserParamsDto> {

    private final UserRepository userRepository;

    public UserParamsMapper(UserRepository userRepository) {
        this.userRepository = userRepository;
    }

    @PostConstruct
    public void init() {
        mapper.createTypeMap(UserParams.class, UserParamsDto.class)
                .addMappings(m -> m.skip(UserParamsDto::setUser)).setPostConverter(toDtoConverter());
        mapper.createTypeMap(UserParamsDto.class, UserParams.class)
                .addMappings(m -> m.skip(UserParams::setUser)).setPostConverter(toEntityConverter());
    }

    @Override
    protected void mapSpecificFields(UserParams source, UserParamsDto destination) {
        whenNotNull(source.getUser(), u -> destination.setUser(u.getId()));
    }

    @Override
    protected void mapSpecificFields(UserParamsDto source, UserParams destination) {
        whenNotNull(source.getUser(), u -> destination.setUser(userRepository.findById(u).orElse(null)));
    }
}

File: src/main/kotlin/ru/xpendence/slimer/mapper/impl/UserParamsMapper.kt
package ru.xpendence.slimer.mapper.impl

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Component
import ru.xpendence.slimer.dto.UserParamsDto
import ru.xpendence.slimer.entity.User
import ru.xpendence.slimer.entity.UserParams
import ru.xpendence.slimer.mapper.AbstractMapper
import ru.xpendence.slimer.mapper.Mapper
import ru.xpendence.slimer.repository.UserRepository
import javax.annotation.PostConstruct

/**
 * Author: Vyacheslav Chernyshov
 * Date: 2019-06-08
 * Time: 23:04
 * e-mail: 2262288@gmail.com
 */
@Component
@Mapper(entity = UserParams::class, dto = UserParamsDto::class)
class UserParamsMapper : AbstractMapper<UserParams, UserParamsDto>() {

    @field:Autowired
    private lateinit var userRepository: UserRepository

    @PostConstruct
    fun init() {
        mapper.createTypeMap(UserParams::class.javaObjectType, UserParamsDto::class.javaObjectType)
                .addMappings { m ->
                    m.skip { destination: UserParamsDto?, source: UserParams? ->
                        destination?.user = source?.user?.id
                    }
//                    m.skip { destination: UserParamsDto?, source: UserParams? ->
//                        destination?.user = source?.user?.id
//                    }
                }
                .postConverter = toDtoConverter()
        mapper.createTypeMap(UserParamsDto::class.javaObjectType, UserParams::class.javaObjectType)
                .addMappings { m ->
                    m.skip { destination: UserParams?, source: UserParamsDto? ->
                        destination?.user = User()
                    }
                }
                .postConverter = toEntityConverter()
    }

    override fun mapSpecificFields(source: UserParams, destination: UserParamsDto) {
        println()
        if (source.user != null) destination.user = source.user!!.id
    }

    override fun mapSpecificFields(source: UserParamsDto, destination: UserParams) {
        if (source.user != null) destination.user = userRepository.findByIdOrNull(source.user)
    }
}
